File: python_solutions/1463.cherry-pickup-ii.cpp
// Containers
#include <deque>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "conf.d/say.h"

using namespace std;

// --------------------------------------------------------
class Solution {
public:
  int dp[70][70][70];
  int dirs[10] = {-1, -1, 0, -1, 1, 0, 0, 1, 1, -1};
  
  int dfs(int R, int C, int r, int c1, int c2, vector<vector<int>> &grid) {
    if (r == R) return 0;
    int &ans = dp[r][c1][c2];
    if (ans > -1) return ans;

    for (int i = 0; i < 9; i++) {
      int nc1 = c1 + dirs[i], nc2 = c2 + dirs[i + 1];
      if (nc1 >= 0 && nc1 <= C - 1 && nc2 >= 0 && nc2 <= C - 1 && nc2 > nc1)
        ans = max(ans, dfs(R, C, r + 1, nc1, nc2, grid));
    }
    ans += grid[r][c1] + grid[r][c2];
    return ans;
  }

  int cherryPickup(vector<vector<int>> &grid) {
    int res = 0;
    int R = grid.size(), C = grid[0].size();
    memset(dp, -1, sizeof(dp));
    return dfs(R, C, 0, 0, C - 1, grid);
  }
};

int main() {
  Solution s;
  vector<vector<int>> grid = {{1, 0, 0, 0, 0, 0, 1},
                              {2, 0, 0, 0, 0, 3, 0},
                              {2, 0, 9, 0, 0, 0, 0},
                              {0, 3, 0, 5, 4, 0, 0},
                              {1, 0, 2, 3, 0, 0, 6}};
  grid = {{8, 8, 10, 9, 1, 7}, {8, 8, 1, 8, 4, 7}, {8, 6, 10, 3, 7, 7}, {3, 0, 9, 3, 2, 7}, {6, 8, 9, 4, 2, 5},
          {1, 1, 5, 8, 8, 1},  {5, 6, 5, 2, 9, 9}, {4, 4, 6, 2, 5, 4},  {4, 4, 5, 3, 1, 6}, {9, 2, 2, 1, 9, 3}};

  say(s.cherryPickup(grid));
}

File: python_solutions/1463.cherry-pickup-ii.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


# https://leetcode.com/problems/cherry-pickup-ii/description/
# Hard
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        dp = defaultdict(lambda: -1)
        R, C = len(grid), len(grid[0])
        dirs = [-1, -1, 0, -1, 1, 0, 0, 1, 1, -1]

        @lru_cache(None)
        def dfs(r, c1, c2):
            if r == R: return 0
            val = dp[(r, c1, c2)]
            if val > -1: return val

            for i in range(9):
                nc1, nc2 = c1 + dirs[i], c2 + dirs[i + 1]
                if 0 <= nc1 <= C - 1 and 0 <= nc2 <= C - 1 and nc1 < nc2:
                    val = max(val, dfs(r + 1, nc1, nc2))
            dp[(r, c1, c2)] = val + grid[r][c1] + grid[r][c2]
            return dp[(r, c1, c2)]

        return dfs(0, 0, C - 1)


gs = [[3, 1, 1], [2, 5, 1], [1, 5, 5], [2, 1, 1]]
print(Solution().cherryPickup(gs))
